src/main/kotlin/com/terraformation/backend/accelerator/ParticipantProjectSpeciesService.kt

```diff
@@ -52,29 +53,42 @@ class ParticipantProjectSpeciesService(
     return dslContext.transactionResult { _ ->
       val existingModels = participantProjectSpeciesStore.create(projectIds, speciesIds)
 
-      val checkedProjectIds = emptySet<ProjectId>()
+      // Used to save relatively expensive queries for projects which we know have submissions
+      val projectDeliverableIds = mutableMapOf<ProjectId, DeliverableId>()
 
-      existingModels.forEach {
-        // A submission must exist for every project that is getting a new species assigned
-        if (checkedProjectIds.contains(it.projectId)) {
+      existingModels.forEach { participantProjectSpecies ->
+        if (projectDeliverableIds.contains(participantProjectSpecies.projectId)) {
+          publishEditEvent(
+              projectDeliverableIds[participantProjectSpecies.projectId]!!,
+              participantProjectSpecies)
```

Comment:
This is totally fine as-is. Two alternate ideas, in case you prefer one of them:
```kotlin
if (participantProjectSpecies.projectId in projectDeliverableIds) { ... }
```
```kotlin
projectDeliverableIds[participantProjectSpecies.projectId]?.let {
  publishEditEvent(it, participantProjectSpecies)
  return@forEach
}
```

Reply:
Thanks, that second one looks like the way to do it in Kotlin